internal/policy/policy.go

```diff
@@ -818,7 +819,122 @@ func ListRules(ctx context.Context, repo *git.Repository, targetRef string) ([]*
+
+	fullDiff.WriteString("Root role keys:\n")
+	fullDiff.WriteString(findDiffBetweenStrings(strings.Join(policyRoot.Roles[RootRoleName].KeyIDs, ", "), strings.Join(policyStagingRoot.Roles[RootRoleName].KeyIDs, ", ")))
+
+	fullDiff.WriteString("\n")
+	for policyPointer < len(policyRules) || policyStagingPointer < len(policyStagingRules) {
+		policyString, policyStagingString := getRuleString(policyRules[policyPointer].Delegation, policyRules[policyPointer].Depth), getRuleString(policyStagingRules[policyStagingPointer].Delegation, policyStagingRules[policyStagingPointer].Depth)
+		switch {
+		case policyPointer == len(policyRules):
+			fullDiff.WriteString(findDiffBetweenStrings("", policyStagingString))
+			policyStagingPointer++
+		case policyStagingPointer == len(policyStagingRules):
+			fullDiff.WriteString(findDiffBetweenStrings(policyString, ""))
+			policyPointer++
+		case policyRules[policyPointer] != policyStagingRules[policyStagingPointer]:
+			switch {
+			case policyRules[policyPointer].Depth < policyStagingRules[policyStagingPointer].Depth:
+				fullDiff.WriteString(findDiffBetweenStrings("", policyStagingString))
+				policyStagingPointer++
+			case policyRules[policyPointer].Depth > policyStagingRules[policyStagingPointer].Depth:
+				fullDiff.WriteString(findDiffBetweenStrings(policyString, ""))
+				policyPointer++
+			default:
+				fullDiff.WriteString(findDiffBetweenStrings(policyString, policyStagingString))
+				policyPointer++
+				policyStagingPointer++
+			}
+		}
+	}
+	return fullDiff.String()
+}
+
+// findDiffBetweenStrings takes two strings (currently only used for
+// rule strings) and using the go-diff/diffmatchpatch library, creates a
+// color coded string of the difference between both strings.
+func findDiffBetweenStrings(initial, withChanges string) string {
+	dmp := diffmatchpatch.New()
+	lines1, lines2, lineArray := dmp.DiffLinesToChars(initial, withChanges)
+	diffs := dmp.DiffMain(lines1, lines2, false)
+	diffs = dmp.DiffCharsToLines(diffs, lineArray)
+
+	var diffDisplay strings.Builder
+
+	for _, diff := range diffs {
+		text := strings.TrimSuffix(diff.Text, "\n")
+		lines := strings.Split(text, "\n")
+
+		for _, line := range lines {
+			switch diff.Type {
+			case diffmatchpatch.DiffInsert:
+				diffDisplay.WriteString(fmt.Sprintf("\033[32m+   %s\033[0m\n", line)) // Green for additions
+			case diffmatchpatch.DiffDelete:
+				diffDisplay.WriteString(fmt.Sprintf("\033[31m-   %s\033[0m\n", line)) // Red for deletions
+			case diffmatchpatch.DiffEqual:
+				diffDisplay.WriteString(fmt.Sprintf("    %s\n", line))
+			}
+		}
+	}
+	return diffDisplay.String()
+}
+
+// getRuleString is a helper that takes a rule, and its depth in the
+// tuf delegation tree, and outputs the rule fields in a user-friendly way
+func getRuleString(rule tuf.Delegation, depth int) string {
+	var changes string
+
+	changes += fmt.Sprintf(strings.Repeat("    ", depth)+"Rule %s:\n", rule.Name)
```

Comment:
Could this be cleaned up a bit so there's not a mix of `+` to concat strings and fmt.Sprintf?

Reply:
Wonder if the indent string at one scope can be computed once too :thinking: